tests/gamma/affinity_session_drain_test.py

```diff
@@ -0,0 +1,253 @@
+        """Load all metadata_keys by default."""
+        if not metadata_keys:
+            metadata_keys = (session_affinity_mixin.COOKIE_METADATA_KEY,)
+        return super().getClientRpcStats(
+            test_client, num_rpcs, metadata_keys=metadata_keys
+        )
+
+    def test_session_drain(self):
+        test_servers: list[server_app.XdsTestServer]
+        # Pod names correspond to test_server hostnames.
+        initial_pods: set[str]
+        with self.subTest("01_run_test_server"):
+            test_servers = self.startTestServers(replica_count=REPLICA_COUNT)
+            initial_pods = set(self.server_runner.pods_started.keys())
+            self.assertLen(initial_pods, REPLICA_COUNT)
+
+        with self.subTest("02_create_ssa_policies"):
+            self.server_runner.create_session_affinity_policy_route()
+            self.server_runner.create_backend_policy(
+                draining_timeout=DRAINING_TIMEOUT,
+            )
+
+        # Default is round-robin LB policy.
+        # TODO(sergiitk): what does it mean? The doc says MAGLEV.
+        # https://cloud.google.com/kubernetes-engine/docs/how-to/configure-gateway-resources#session_affinity
+
+        test_client: client_app.XdsTestClient
+        with self.subTest("03_start_test_client"):
+            test_client = self.startTestClient(test_servers[0])
+
+        with self.subTest("04_confirm_all_servers_receive_traffic"):
+            # Note: the output of this call may or may not print out the
+            # cookie. This is *not* any kind of signal, just depends on system
+            # latency. This subtest does not use any cookies.
+            self.assertRpcsEventuallyGoToGivenServers(
+                test_client,
+                test_servers,
+                num_rpcs=120,  # Nice and even.
+                retry_timeout=TRAFFIC_PIN_TIMEOUT,
+                retry_wait=TRAFFIC_PIN_RETRY_WAIT,
+            )
+            logger.info(
+                "Confirmed all servers received traffic: %s",
+                [server.hostname for server in test_servers],
+            )
+
+        with self.subTest("05_retrieve_cookie"):
+            result = self.assertSsaCookieAssigned(test_client, test_servers)
+            cookie: Final[str] = result[0]
+            chosen_server: Final[server_app.XdsTestServer] = result[1]
+            # The name of the chosen server, same as the pod name.
+            chosen_name: Final[str] = chosen_server.hostname
+            logger.info("Chosen server: %s, cookie: %s", chosen_name, cookie)
+
+        with self.subTest("06_only_chosen_server_receives_rpcs_with_cookie"):
+            logger.info("Configuring client to send cookie %s", cookie)
+            test_client.update_config.configure_unary(
+                metadata=(
+                    (grpc_testing.RPC_TYPE_UNARY_CALL, "cookie", cookie),
+                ),
+            )
+            self.assertRpcsEventuallyGoToGivenServers(
```

Comment:
Do we need check `num_rpcs` here?

Reply:
Not sure what you mean by that

Reply:
Discussed IRL, default `num_rpcs=100` works ok here, because assert it's all sent to the single server only.